cordapp/src/main/kotlin/net/corda/plugins/ByteArrayStuff.kt

```diff
@@ -0,0 +1,284 @@
+package net.corda.plugins
+
+import java.io.ByteArrayInputStream
+import java.io.OutputStream
+import java.nio.ByteBuffer
+import java.security.MessageDigest
+import java.util.*
+import javax.xml.bind.DatatypeConverter
+
+open class OpaqueBytes(bytes: ByteArray) : ByteSequence(bytes, 0, bytes.size) {
+    companion object {
+        /**
+         * Create [OpaqueBytes] from a sequence of [Byte] values.
+         */
+        @JvmStatic
+        fun of(vararg b: Byte) = OpaqueBytes(byteArrayOf(*b))
+    }
+
+    init {
+        require(bytes.isNotEmpty()) { "Byte Array must not be empty" }
+    }
+
+    /**
+     * The bytes are always cloned so that this object becomes immutable. This has been done
+     * to prevent tampering with entities such as [net.corda.core.crypto.SecureHash] and [net.corda.core.contracts.PrivacySalt], as well as
+     * preserve the integrity of our hash constants [net.corda.core.crypto.SecureHash.zeroHash] and [net.corda.core.crypto.SecureHash.allOnesHash].
+     *
+     * Cloning like this may become a performance issue, depending on whether or not the JIT
+     * compiler is ever able to optimise away the clone. In which case we may need to revisit
+     * this later.
+     */
+    final override val bytes: ByteArray = bytes
+        get() = field.clone()
+}
+
+sealed class ByteSequence(private val _bytes: ByteArray, val offset: Int, val size: Int) : Comparable<ByteSequence> {
```

Comment:
Why not use `ByteBuffer`? (This code has been copied and pasted from Corda, hasn't it... :wink:)